### DISCUSSION TITLE
Static build for cpanel and other hostings

### DISCUSSION THREAD

**Discussion by NEXTink** [vercel/next.js#49385] [Help]
### Summary

I am trying to host my site on a panel. Everytime I use next build it returns CSS and js but not HTML pages and my server cannot interpret that.

### Example
It justs returns an out folder that contains an _next folder which contains some CSS files and js files

_No response_

**Comment by chrisjcthomas** [comment]
For cPanel/static hosting, you need a real static export, not the normal server build.

In current Next.js, that means adding this to `next.config.js`:

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  output: 'export',
}

module.exports = nextConfig
```

Then run:

```bash
next build
```

That should produce an `out/` folder containing HTML files plus the `_next/` assets. Upload the *contents of `out/`* to the cPanel public directory.

If you only see `_next` assets and no useful HTML pages, I would check these first:

- make sure `output: 'export'` is actually in the config being used by the build
- make sure your routes can be statically generated
- for dynamic routes, provide static params (`generateStaticParams` in the App Router)
- avoid server-only features on this deployment target

Static export will not support things that require a Next.js server, such as API routes, middleware, server actions, ISR, or the default image optimizer. If the app depends on those, cPanel-style static hosting is the wrong target; you would need Node hosting or a platform that runs the Next server.

**Comment by Hassanmahmood4** [comment]
```markdown
If your `out` folder only contains `_next` (CSS/JS assets) and no `.html` files, the build isn't actually generating a static export. A few things to check:

**1. Make sure static export is enabled in `next.config.js`:**

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  output: 'export',
};

module.exports = nextConfig;
```

Without `output: 'export'`, `next build` produces a server build in `.next` — which is meant to be run with `next start` on a Node server, not uploaded to a static hosting panel. (The old `next export` command was removed in Next.js 14 — this config option replaces it.)

**2. Check you're uploading the right folder.**
After building with `output: 'export'`, the deployable folder is `out/`, and it should contain `index.html` plus one `.html` file per route at the top level:

```
out/
├── index.html
├── about.html
├── _next/
│   └── static assets...
```

If you see HTML files there, upload the **contents** of `out/` to your panel's public_html (not the folder itself nested inside).

**3. If the build errors or skips pages**, you're probably using features that can't be statically exported:
- API routes (`app/api/*` or `pages/api/*`)
- Server-side rendering (`getServerSideProps`, `cookies()`, `headers()`)
- Dynamic routes without `generateStaticParams` (App Router) or `getStaticPaths` (Pages Router)
- Default `next/image` loader — set `images: { unoptimized: true }` in the config

For dynamic routes, add:

```js
// app/blog/[slug]/page.js
export function generateStaticParams() {
  return [{ slug: 'post-1' }, { slug: 'post-2' }];
}
```

**4. Clean rebuild** after changing the config:

```bash
rm -rf .next out
npm run build
```

Then check `out/` — you should now see your HTML pages alongside `_next/`.
```

The most likely cause for their exact symptom is **#1** — they're missing `output: 'export'` and uploading a server build to a static host.
